src/object/obj_rpc.c

```diff
@@ -644,12 +627,625 @@ crt_proc_struct_daos_shard_tgt(crt_proc_t proc, struct daos_shard_tgt *st)
+			return -DER_NOMEM;
+	}
+
+	for (i = 0; i < count; i++) {
+		rc = crt_proc_struct_daos_cpd_req_idx(proc,
+						      &dcde->dcde_reqs[i]);
+		if (rc != 0) {
+			if (proc_op == CRT_PROC_DECODE)
+				D_FREE(dcde->dcde_reqs);
+
+			return -DER_HG;
+		}
+	}
+
+	return 0;
+}
+
+static int
+crt_proc_struct_daos_cpd_sg(crt_proc_t proc, struct daos_cpd_sg *dcs)
+{
+	crt_proc_op_t	proc_op;
+	int		rc;
+	int		i;
+
+	rc = crt_proc_uint32_t(proc, &dcs->dcs_type);
+	if (rc != 0)
+		return -DER_HG;
+
+	rc = crt_proc_uint32_t(proc, &dcs->dcs_nr);
+	if (rc != 0)
+		return -DER_HG;
+
+	rc = crt_proc_get_op(proc, &proc_op);
+	if (rc != 0)
+		return -DER_HG;
+
+	switch (dcs->dcs_type) {
+	case DCST_HEAD: {
+		struct daos_cpd_sub_head	*dcsh;
+
+		if (proc_op == CRT_PROC_DECODE) {
+			D_ALLOC_ARRAY(dcsh, dcs->dcs_nr);
+			if (dcsh == NULL)
+				D_GOTO(out, rc = -DER_NOMEM);
+
+			dcs->dcs_buf = dcsh;
+		} else {
+			dcsh = dcs->dcs_buf;
+		}
+
+		for (i = 0; i < dcs->dcs_nr; i++) {
+			rc = crt_proc_struct_daos_cpd_sub_head(proc, &dcsh[i]);
+			if (rc != 0)
+				D_GOTO(out, rc = -DER_HG);
+		}
+
+		break;
+	}
+	case DCST_REQ_CLI:
+	case DCST_REQ_SRV: {
+		struct daos_cpd_sub_req		*dcsr;
+		bool				 with_oid;
+
+		if (proc_op == CRT_PROC_DECODE) {
+			D_ALLOC_ARRAY(dcsr, dcs->dcs_nr);
+			if (dcsr == NULL)
+				D_GOTO(out, rc = -DER_NOMEM);
+
+			dcs->dcs_buf = dcsr;
+			with_oid = true;
+		} else {
+			dcsr = dcs->dcs_buf;
+			if (dcs->dcs_type == DCST_REQ_SRV)
+				with_oid = true;
+			else
+				with_oid = false;
+		}
+
+		for (i = 0; i < dcs->dcs_nr; i++) {
+			rc = crt_proc_struct_daos_cpd_sub_req(proc, &dcsr[i],
+							      with_oid);
+			if (rc != 0)
+				D_GOTO(out, rc = -DER_HG);
+		}
+
+		break;
+	}
+	case DCST_DISP: {
+		struct daos_cpd_disp_ent	*dcde;
+
+		if (proc_op == CRT_PROC_DECODE) {
+			D_ALLOC_ARRAY(dcde, dcs->dcs_nr);
+			if (dcde == NULL)
+				D_GOTO(out, rc = -DER_NOMEM);
+
+			dcs->dcs_buf = dcde;
+		} else {
+			dcde = dcs->dcs_buf;
+		}
+
+		for (i = 0; i < dcs->dcs_nr; i++) {
+			rc = crt_proc_struct_daos_cpd_disp_ent(proc, &dcde[i]);
+			if (rc != 0)
+				D_GOTO(out, rc = -DER_HG);
+		}
+
+		break;
+	}
+	case DCST_TGT: {
+		struct daos_shard_tgt		*dst;
+
+		if (proc_op == CRT_PROC_DECODE) {
+			D_ALLOC_ARRAY(dst, dcs->dcs_nr);
+			if (dst == NULL)
+				D_GOTO(out, rc = -DER_NOMEM);
+
+			dcs->dcs_buf = dst;
+		} else {
+			dst = dcs->dcs_buf;
+		}
+
+		for (i = 0; i < dcs->dcs_nr; i++) {
+			rc = crt_proc_struct_daos_shard_tgt(proc, &dst[i]);
+			if (rc != 0)
+				D_GOTO(out, rc = -DER_HG);
+		}
+
+		break;
+	}
+	default:
+		return -DER_INVAL;
+	}
+
+out:
+	/* XXX: There is potential memory leak for the case of CRT_PROC_DECODE
+	 *	with failure. We may allocate some DRAM in some low layer proc
+	 *	functions when decoding former elements.
+	 *
+	 *	Currently, We seems not have efficient way to release them. It
+	 *	is not special for CPD related proc interfaces, instead, it is
+	 *	general issue for the whole CRT proc mechanism.
```

Comment:
Not sure if I understood it. For example in proc_A() it calls proc_B(). 
When DECOCE, the proc_B() allocated some memory, assume it should assign the allocated memory pointer to somewhere in data-structure aware by proc_A(), then if in proc_A hits some other failure it has the chance to free mem allocated in proc_B()?

Reply:
If proc_B() allocate some DRAM and attach it under the sub-level of the proc_A's data structure, then if proc_A failed, we may have no change to free such allocated DRAM.